omen2/omen.py

```diff
@@ -0,0 +1,214 @@
+        return self[self.table_types[table_name]]
+
+    def __getitem__(self, table_type: Type[T]) -> T:
+        return self.tables[table_type]
+
+    def __setitem__(self, table_type: Type[T], table: T):
+        assert table_type.table_name == table.table_name
+        self.table_types[table.table_name] = table_type
+        self.validate_table(table.table_name, table_type)
+        self.tables[table_type] = table
+
+    def set_table(self, table: Table):
+        self[type(table)] = table
+
+    def load_dict(self, data_set: Dict[str, Iterable[Dict[str, Any]]]):
+        # load sample data into self
+        for name, values in data_set.items():
+            tab: Table = self.get_table_by_name(name)
+            for entry in values:
+                tab.add(tab.row_type._from_db(entry))
+
+    def dump_dict(self) -> Dict[str, Iterable[Dict[str, Any]]]:
+        ret = {}
+        for ttype, tab in self.tables.items():
+            lst = []
+            for obj in tab:
+                lst.append(obj._to_db())
+            ret[ttype.table_name] = lst
+        return ret
+
+    def validate_table(self, name, tab):
+        assert issubclass(tab, Table)
+
+        if not getattr(tab, "table_name", None):
+            tab.table_name = name
+        assert tab.table_name == name
+
+        assert issubclass(tab.row_type, ObjBase)
+        assert isinstance(getattr(tab.row_type, "_pk"), tuple)
+        assert issubclass(tab.row_type._table_type, tab)
+
+        if not getattr(tab, "field_names", None):
+            log.debug("%s: default serialization field names used", name)
+            tab.field_names = {c.name for c in self.model[name].columns}
+        assert isinstance(tab.field_names, set)
+
+        pk = None
+        model = self.model[name]
+        for idx in model.indexes:
+            if idx.primary:
+                pk = idx.fields
+        if len(pk) == 1:
+            for fd in model.columns:
+                if fd.name == pk[0]:
+                    tab.allow_auto = fd.autoinc
+
+    @classmethod
+    def codegen(cls, force=False):
+        """Generate code derived from my model, and put it next to my __file__."""
+        if cls.__module__ == "__main__":
+            module, _ = os.path.splitext(
+                os.path.basename(sys.modules["__main__"].__file__)
+            )
+        else:
+            module = cls.__module__
+
+        module += "_gen"
+
+        try:
+            if force:
+                raise ImportError
+            generated = importlib.import_module(module)
+        except (ImportError, SyntaxError):
+            CodeGen.generate_from_class(cls)
+            generated = importlib.import_module(module)
+
+        for name in generated.__all__:
+            table_type = getattr(generated, name)
+            if name not in cls.table_types:
+                cls.table_types[name] = table_type
+
+    @staticmethod
+    def __multi_query(db, sql):
+        unlikely = "@!~@"
```

Comment:
Might be worth adding an assert like `assert unlikely not in sql`, even if it's, well, unlikely

Reply:
done